processing/src/main/java/io/druid/segment/data/CompressedVSizeIntsIndexedSupplier.java

```diff
@@ -0,0 +1,393 @@
+
+                  @Override
+                  public ResourceHolder<ByteBuffer> next()
+                  {
+                    ByteBuffer retVal = ByteBuffer
+                        .allocate(chunkBytes)
+                        .order(byteOrder);
+
+                    if (chunkFactor > list.size() - position) {
+                      retVal.limit((list.size() - position) * numBytes);
+                    } else {
+                      retVal.limit(chunkFactor * numBytes);
+                    }
+
+                    final List<Integer> ints = list.subList(position, position + retVal.remaining() / numBytes);
+                    for(int value : ints) {
+                      if(byteOrder.equals(ByteOrder.BIG_ENDIAN)) {
+                        byte[] intAsBytes = Ints.toByteArray(value);
+                        retVal.put(intAsBytes, intAsBytes.length - numBytes, numBytes);
+                      } else {
+                        final ByteBuffer buf = ByteBuffer
+                            .allocate(Ints.BYTES)
+                            .order(byteOrder)
+                            .putInt(value);
+                        retVal.put(buf.array(), 0, numBytes);
+                      }
+                    }
+                    retVal.rewind();
+                    position += retVal.remaining() / numBytes;
+
+                    return StupidResourceHolder.create(retVal);
+                  }
+
+                  @Override
+                  public void remove()
+                  {
+                    throw new UnsupportedOperationException();
+                  }
+                };
+              }
+            },
+            CompressedByteBufferObjectStrategy.getBufferForOrder(byteOrder, compression, chunkBytes)
+        ),
+        compression
+    );
+  }
+
+  private class CompressedFullSizeIndexedInts extends CompressedVSizeIndexedInts {
+    IntBuffer intBuffer;
+
+    @Override
+    protected void loadBuffer(int bufferNum)
+    {
+      super.loadBuffer(bufferNum);
+      intBuffer = buffer.asIntBuffer();
+    }
+
+    @Override
+    protected int _get(int index)
+    {
+      return intBuffer.get(intBuffer.position() + index);
+    }
+  }
+
+  private class CompressedShortSizeIndexedInts extends CompressedVSizeIndexedInts {
+    ShortBuffer shortBuffer;
+
+    @Override
+    protected void loadBuffer(int bufferNum)
+    {
+      super.loadBuffer(bufferNum);
+      shortBuffer = buffer.asShortBuffer();
+    }
+
+    @Override
+    protected int _get(int index)
+    {
+      // removes the need for padding
+      return shortBuffer.get(shortBuffer.position() + index) & 0xFFFF;
+    }
+  }
+
+  private class CompressedByteSizeIndexedInts extends CompressedVSizeIndexedInts {
+    @Override
+    protected int _get(int index)
+    {
+      // removes the need for padding
+      return buffer.get(buffer.position() + index) & 0xFF;
+    }
+  }
+
+  private class CompressedVSizeIndexedInts implements IndexedInts
```

Comment:
yes, it implements the general case for any numBytes. Others are just optimizations.